File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    f_dic = {}
    sum_dic = {}
    answer = 0

    # for

    for i in range(len(friends)):
        f_dic[friends[i]] = {}
        sum_dic[friends[i]] = [0, 0]
        for j in range(len(friends)):
            if i == j:
                continue
            f_dic[friends[i]][friends[j]] = 0

    # print(f_dic)
    dic = {}

    for i in range(len(gifts)):
        a = gifts[i].split()
        send, receive = a[0], a[1]
        f_dic[send][receive] += 1

        if send not in sum_dic:
            sum_dic[send] = [1, 0]
        else:
            sum_dic[send][0] += 1

        if receive not in sum_dic:
            sum_dic[receive] = [0, 1]
        else:
            sum_dic[receive][1] += 1

    # print(sum_dic)
    # return 0

    for key_1, value_1 in f_dic.items():
        # a = sorted(value_1.values(), key=lambda x: x)
        # if max(a) == 0:
        #     continue

        dic[key_1] = 0
        for key_2, value_2 in value_1.items():
            key_2_to_key_1 = f_dic[key_2][key_1]
            if value_2 > key_2_to_key_1:
                dic[key_1] += 1
            elif key_2_to_key_1 > value_2:
                continue
            elif key_2_to_key_1 == value_2 or value_2 == 0:
                if key_2 not in sum_dic:
                    continue
                elif key_1 not in sum_dic:
                    continue
                s_1 = sum_dic[key_1][0] - sum_dic[key_1][1]
                s_2 = sum_dic[key_2][0] - sum_dic[key_2][1]

                if s_1 > s_2:
                    dic[key_1] += 1

    answer = max(sorted(dic.values(), key=lambda x: x))
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
//     두 사람이 선물받은 기록 잇다 => 더 많이 준 사람이 선물 받음
//     두 사람이 기록이 없거나 주고받은 회수가 같다 => 선물 지수가 큰 사람이 선물 받음
//     선물 지수는 이번달까지 자신이 친구들에게 준 선물의 수 - 받은 선물의 수
//     선물지수도 같다면 주고받지 않음
    let obj = {}
    let cnt = {}
    for (let f of friends){
            obj[f] = [0,{}];
            cnt[f] = 0
    }
    for(let g of gifts){
        const [giver, rec] = g.split(' ');
        obj[giver][0] += 1;
        if(obj[giver][1][rec]){
            obj[giver][1][rec] += 1
        }else{
            obj[giver][1][rec] = 1
        }
        obj[rec][0] -= 1;
        if(obj[rec][1][giver]){
            obj[rec][1][giver] -= 1
        }else{
            obj[rec][1][giver] = -1
        }
    }
    
    
    for(let gi of friends){
        for(let gi2 of friends){
            if(gi === gi2) continue;
            if(obj[gi][1][gi2] > obj[gi2][1][gi]){
                cnt[gi] += 1
            }else if (obj[gi][1][gi2] === obj[gi2][1][gi] ){
                if(obj[gi][0] > obj[gi2][0]){
                    cnt[gi] += 1
                }
            }
        }
    }
    const arr = Object.entries(cnt);
    arr.sort((a,b)=> b[1]-a[1]);
    return arr[0][1]
}